examples/example_uq_bootstrap.py

```diff
@@ -0,0 +1,131 @@
+"""
+.. _tut_bootstrap:
+
+Bootstrapping
+=============
+
+In this example, we demonstrate how to perform uncertainty quantification (UQ) using
+bootstrap method. We use a Stillinger-Weber (SW) potential for silicon that is archived
+in OpenKIM_.
+
+For simplicity, we only set the energy-scaling parameters, i.e., ``A`` and ``lambda`` as
+the tunable parameters. These parameters will be calibrated to energies and forces of a
+small dataset, consisting of 4 compressed and stretched configurations of diamond silicon
+structure.
+"""
+
+
+##########################################################################################
+# To start, let's first install the SW model::
+#
+#    $ kim-api-collections-management install user SW_StillingerWeber_1985_Si__MO_405512056662_006
+#
+# .. seealso::
+#    This installs the model and its driver into the ``User Collection``. See
+#    :ref:`install_model` for more information about installing KIM models.
+
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+from kliff.calculators import Calculator
+from kliff.dataset import Dataset
+from kliff.loss import Loss
+from kliff.models import KIMModel
+from kliff.models.parameter_transform import LogParameterTransform
+from kliff.uq.bootstrap import BootstrapEmpiricalModel
+from kliff.utils import download_dataset
+
+##########################################################################################
+# Before running bootstrap, we need to define a loss function and train the model. More
+# detail information about this step can be found in :ref:`tut_kim_sw` and
+# :ref:`tut_params_transform`.
+
+# Instantiate a transformation class to do the log parameter transform
+param_names = ["A", "lambda"]
+
+# Create the model
+model = KIMModel(model_name="SW_StillingerWeber_1985_Si__MO_405512056662_006")
+
+# Set the tunable parameters and the initial guess
+opt_params = {"A": [["default"]], "lambda": [["default"]]}
+
+model.set_opt_params(**opt_params)
+model.echo_opt_params()
+
+# Get the dataset
+dataset_path = download_dataset(dataset_name="Si_training_set_4_configs")
+# Read the dataset
+tset = Dataset(dataset_path)
+configs = tset.get_configs()
+
+# Create calculator
+calc = Calculator(model)
+# Only use the forces data
+ca = calc.create(configs, use_energy=False, use_forces=True)
+
+# Instantiate the loss function
+residual_data = {"normalize_by_natoms": False}
+loss = Loss(calc, residual_data=residual_data)
+
+# Train the model
+min_kwargs = dict(method="lm")  # Optimizer setting
+loss.minimize(**min_kwargs)
```

Comment:
We don't necessarily need to train against the entire dataset before bootstrapping. It is ok to do this, but typical bootstrapping does not do this, right? 

This can actually lead to data leakage because the model is first trained on the entire dataset.

Reply:
Let's first make sure we are on the same footing. Below is a general way to do bootstrapping:
1. sample training set with replacement 
2. train the model with the sampled data 
3. and redo 1 and 2 multiple times 
4. evaluate on a holdout test set (mean and stdev mostly)

You see here we do not have the training on entire training set step. It does not have anything to do with running time, but how we need to do it. My argument was that we don't need to do the training on the entire training set step. It might be ok to do it, but it is then not the typical bootstrapping approach. 

Also, I don't think, in principle, the optimizations in different runs need to use the same optimizing hyerparams, although it is typically the case.

Reply:
> Let's first make sure we are on the same footing. Below is a general way to do bootstrapping:
> 
> 1. sample training set with replacement
> 2. train the model with the sampled data
> 3. and redo 1 and 2 multiple times
> 4. evaluate on a holdout test set (mean and stdev mostly)

Yes, this is the general way to do bootstrapping.

> You see here we do not have the training on entire training set step. It does not have anything to do with running time, but how we need to do it. My argument was that we don't need to do the training on the entire training set step. It might be ok to do it, but it is then not the typical bootstrapping approach.
> 
> Also, I don't think, in principle, the optimizations in different runs need to use the same optimizing hyerparams, although it is typically the case.

I see what you meant. I will argue that we need to keep the same optimizing hyperparameters. We can think that there is a mapping from the bootstrap dataset to the inferred parameters. For each run, we suggest using the same mapping, i.e., the same optimizing hyperparameters. Otherwise, the mapping from the bootstrap dataset to the parameters is different for each run, and the inferred parameters cannot be compared directly.

In the example, I trained the potential before running bootstrapping to set the optimizing hyperparameters to use in the bootstrapping. I guess, this step (training to the original dataset) is not necessary, but using the same optimizer setting is an important step in bootstrapping.